section/Lua/TableFuncs.cpp

```diff
@@ -0,0 +1,177 @@
+    uint8_t lsizenode = table->lsizenode;
+    if (lsizenode)
+    {
+        uint32_t num_nodes = 1 << lsizenode;
+        auto node = table->node;
+        do
+        {
+            if (node->i_val.tt)
+                ++size;
+            ++node;
+            --num_nodes;
+        } while (num_nodes);
+    }
+    int sizearray = table->sizearray;
+    if (sizearray)
+    {
+        auto array = table->array;
+        do
+        {
+            if (array->tt)
+                ++size;
+            ++array;
+            --sizearray;
+        } while (sizearray);
+    }
+
+    lua_pushnumber(L, size);
+    return 1;
+}
+
+int lua_tableempty(lua_State *L)
+{
+    auto base = GetField<lua::TObject *>(L, 0xC);
+    auto top = GetField<lua::TObject *>(L, 0x8);
+    if (base >= top || base->tt != LUA_TTABLE)
+    {
+        lua_pushboolean(L, true);
+        return 1;
+    }
+
+    auto table = (lua::Table *)base->value;
+    constexpr auto has_nodes = [](lua::Table *t)
+    {
+        uint8_t lsizenode = t->lsizenode;
+        if (!lsizenode)
+            return false;
+
+        uint32_t num_nodes = 1 << lsizenode;
+        auto node = t->node;
+        while (node->i_val.tt == LUA_TNIL)
+        {
+            ++node;
+            if (!--num_nodes)
+                return false;
+        }
+        return true;
+    };
+
+    constexpr auto has_array = [](lua::Table *t)
+    {
+        int sizearray = t->sizearray;
+        if (!sizearray)
+            return false;
+
+        auto array = t->array;
+        while (array->tt == LUA_TNIL)
+        {
+            ++array;
+            if (!--sizearray)
+                return false;
+        }
+        return true;
+    };
+
+    lua_pushboolean(L, !(has_nodes(table) || has_array(table)));
+    return 1;
+}
+
+int TableClone(lua_State *L)
+{
+    LuaObject obj{L->LuaState, 1};
+    LuaObject cloned{};
+    obj.Clone(&cloned);
+    cloned.PushStack(L);
+    return 1;
+}
+
+// UI_Lua reprsl({table.unpack2({1,2,3,4},2,3)})
+// UI_Lua reprsl({table.unpack2({1,2,3,4},2)})
+// UI_Lua reprsl({table.unpack2({1,2,3,4})})
+// UI_Lua reprsl({table.unpack2({1,2,3,4},-1)})
+// UI_Lua reprsl({table.unpack2({1,2,3,4},-1000)})
+// UI_Lua reprsl({table.unpack2({1,2,3,4},0, 1000000)}) //stack overflow
+
+int lua_unpack(lua_State *l)
+{
+    luaL_checktype(l, 1, LUA_TTABLE);
+    const int n = lua_getn(l, 1);
+    const int start_i = luaL_optnumber(l, 2, 1);
+    const int end_i = luaL_optnumber(l, 3, n);
+    if (start_i > end_i)
+        return 0;
+
+    const int n_stack = end_i - start_i + 1;
+    luaL_checkstack(l, n_stack, "too many results to unpack");
+    for (int i = start_i; i <= end_i; i++)
+    {
+        lua_rawgeti(l, 1, i);
+    }
+    return n_stack;
+}
+
+const luaL_reg RegTableFuncsDesc[] = {{"getsize2", &lua_tablesize},
+                                      {"empty2", &lua_tableempty},
+                                      {"getn2", (lua_CFunction)0x00927C20},
```

Comment:
Do we still need these `2` alternatives?

Reply:
I can remove them.

Reply:
The problem is Lua resets them in current dev branch:
```lua
--- table.empty(t) returns true iff t has no keys/values.
---@param t table
---@return boolean
function table.empty(t)
    if type(t) ~= 'table' then return true end
    return next(t) == nil
end

--- Returns actual size of a table, including string keys
---@param t table
---@return number
function table.getsize(t)
    if type(t) ~= 'table' then return 0 end
    local size = 0
    for k, v in t do
        size = size + 1
    end
    return size
end

-- replace with assembly implementations
table.empty = table.empty2 or table.empty
table.getsize = table.getsize2 or table.getsize

```

Reply:
Then we remove that logic. As long as the assembly versions are functionally equivalent then that should be no issue.

Reply:
See also: https://github.com/FAForever/fa/pull/6537

Reply:
Done